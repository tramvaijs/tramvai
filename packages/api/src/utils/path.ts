import path from 'node:path';

export const resolveAbsolutePathForFile = ({
  file,
  sourceDir,
  rootDir,
}: {
  file: string;
  sourceDir: string;
  rootDir: string;
}): string => {
  if (path.isAbsolute(file)) {
    return file;
  }
  if (path.isAbsolute(sourceDir)) {
    return path.join(sourceDir, file);
  }
  if (path.isAbsolute(rootDir)) {
    return path.join(rootDir, sourceDir, file);
  }
  return path.resolve(rootDir, sourceDir, file);
};

export const resolveAbsolutePathForFolder = ({
  folder,
  rootDir,
}: {
  folder: string;
  rootDir: string;
}): string => {
  if (path.isAbsolute(folder)) {
    return folder;
  }
  if (path.isAbsolute(rootDir)) {
    return path.join(rootDir, folder);
  }
  return path.resolve(rootDir, folder);
};
