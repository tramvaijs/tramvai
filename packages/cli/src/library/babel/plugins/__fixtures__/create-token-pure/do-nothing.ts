// @ts-ignore
// eslint-disable-next-line import/no-unresolved
import { createApp } from '@tramvai/core';
import { Container } from '@tinkoff/dippy';

const di = new Container();

createApp({
  name: 'test'
});
