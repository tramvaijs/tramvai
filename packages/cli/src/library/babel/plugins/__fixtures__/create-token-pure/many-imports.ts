// @ts-ignore
// eslint-disable-next-line import/no-unresolved
import { createToken } from '@tinkoff/dippy';
import { Module } from '@tramvai/core'

export const TOKEN = createToken('token');
const module = Module({
  providers: []
});
