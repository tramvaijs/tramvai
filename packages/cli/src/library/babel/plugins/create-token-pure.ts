import annotateAsPure from '@babel/helper-annotate-as-pure';
import type { Plugin } from './types.h';

interface InnerState {
  hasCreateToken: boolean;
}

export const createTokenPurePlugin: Plugin<InnerState> = ({ types: t }) => {
  return {
    pre() {
      this.hasCreateToken = false;
    },
    visitor: {
      ImportDeclaration(path) {
        if (this.hasCreateToken) {
          return;
        }

        const { node } = path;
        const sourceValue = node.source.value;

        if (sourceValue === '@tramvai/core' || sourceValue === '@tinkoff/dippy') {
          this.hasCreateToken = node.specifiers.some(
            (s) => s.type === 'ImportSpecifier' && s.local?.name === 'createToken'
          );
        }
      },
      CallExpression(path) {
        const { node } = path;

        if (!this.hasCreateToken) {
          return;
        }

        const { callee } = node;

        if (callee?.type === 'Identifier') {
          if (callee?.name === 'createToken') {
            annotateAsPure(path);
          }
        }
      },
    },
  };
};

export default createTokenPurePlugin;
