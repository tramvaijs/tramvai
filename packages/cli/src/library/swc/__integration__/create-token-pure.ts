import { createToken } from '@tramvai/core';

export const TEST_TOKEN = createToken<string>('TEST_TOKEN');
