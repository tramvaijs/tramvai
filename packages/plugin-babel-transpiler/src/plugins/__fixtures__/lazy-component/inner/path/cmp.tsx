import React from 'react';

const InnerCmp = () => {
  return <div>inner cmp</div>;
}

export default InnerCmp;
