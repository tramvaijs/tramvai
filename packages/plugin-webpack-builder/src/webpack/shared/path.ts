import path from 'node:path';

export const resolveAbsolutePathForEntry = ({
  entry,
  sourceDir,
  rootDir,
}: {
  entry: string;
  sourceDir: string;
  rootDir: string;
}): string => {
  if (path.isAbsolute(entry)) {
    return entry;
  }
  if (path.isAbsolute(sourceDir)) {
    return path.join(sourceDir, entry);
  }
  if (path.isAbsolute(rootDir)) {
    return path.join(rootDir, sourceDir, entry);
  }
  return path.resolve(rootDir, sourceDir, entry);
};

export const resolveAbsolutePathForOutput = ({
  output,
  rootDir,
}: {
  output: string;
  rootDir: string;
}): string => {
  if (path.isAbsolute(output)) {
    return output;
  }
  if (path.isAbsolute(rootDir)) {
    return path.join(rootDir, output);
  }
  return path.resolve(rootDir, output);
};
