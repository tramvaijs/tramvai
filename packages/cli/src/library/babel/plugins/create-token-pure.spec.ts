import path from 'path';
import pluginTester from 'babel-plugin-tester';
import { createTokenPurePlugin } from './create-token-pure';
import babelConfig from '../index';

pluginTester({
  plugin: createTokenPurePlugin,
  pluginName: 'create-token-pure',
  filename: path.join(__dirname, '__fixtures__', 'create-token-pure', 'test'),
  babelOptions: babelConfig({
    typescript: true,
    generateDataQaTag: false,
    loader: false,
    // @ts-expect-error
    markCreateTokenAsPure: false,
  }),
  tests: {
    'do nothing': {
      fixture: 'do-nothing.ts',
      snapshot: true,
    },
    'add pure annotation when createToken from @tramvai/core': {
      fixture: 'from-core.ts',
      snapshot: true,
    },
    'add pure annotation when createToken from @tinkoff/dippy': {
      fixture: 'from-dippy.ts',
      snapshot: true,
    },
    'do not annotate more again': {
      fixture: 'already-has-comment.ts',
      snapshot: true,
    },
    'do not miss createToken on many imports': {
      fixture: 'many-imports.ts',
      snapshot: true,
    },
  },
});
