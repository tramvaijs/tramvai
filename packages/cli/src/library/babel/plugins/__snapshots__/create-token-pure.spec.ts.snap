// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create-token-pure add pure annotation when createToken from @tinkoff/dippy: add pure annotation when createToken from @tinkoff/dippy 1`] = `

// @ts-ignore
// eslint-disable-next-line import/no-unresolved
import { createToken } from '@tinkoff/dippy';

export const TOKEN = createToken('token');

      ↓ ↓ ↓ ↓ ↓ ↓

// @ts-ignore
// eslint-disable-next-line import/no-unresolved
import { createToken } from '@tinkoff/dippy';
export var TOKEN = /*#__PURE__*/ createToken('token');


`;

exports[`create-token-pure add pure annotation when createToken from @tramvai/core: add pure annotation when createToken from @tramvai/core 1`] = `

// @ts-ignore
// eslint-disable-next-line import/no-unresolved
import { createToken } from '@tramvai/core';

export const TOKEN = createToken('token');

      ↓ ↓ ↓ ↓ ↓ ↓

// @ts-ignore
// eslint-disable-next-line import/no-unresolved
import { createToken } from '@tramvai/core';
export var TOKEN = /*#__PURE__*/ createToken('token');


`;

exports[`create-token-pure do not annotate more again: do not annotate more again 1`] = `

// @ts-ignore
// eslint-disable-next-line import/no-unresolved
import { createToken } from '@tinkoff/dippy';

export const TOKEN = /* #__PURE__*/ createToken('token');

      ↓ ↓ ↓ ↓ ↓ ↓

// @ts-ignore
// eslint-disable-next-line import/no-unresolved
import { createToken } from '@tinkoff/dippy';
export var TOKEN = /* #__PURE__*/ createToken('token');


`;

exports[`create-token-pure do not miss createToken on many imports: do not miss createToken on many imports 1`] = `

// @ts-ignore
// eslint-disable-next-line import/no-unresolved
import { createToken } from '@tinkoff/dippy';
import { Module } from '@tramvai/core'

export const TOKEN = createToken('token');
const module = Module({
  providers: []
});

      ↓ ↓ ↓ ↓ ↓ ↓

// @ts-ignore
// eslint-disable-next-line import/no-unresolved
import { createToken } from '@tinkoff/dippy';
import { Module } from '@tramvai/core';
export var TOKEN = /*#__PURE__*/ createToken('token');
var module = Module({
  providers: [],
});


`;

exports[`create-token-pure do nothing: do nothing 1`] = `

// @ts-ignore
// eslint-disable-next-line import/no-unresolved
import { createApp } from '@tramvai/core';
import { Container } from '@tinkoff/dippy';

const di = new Container();

createApp({
  name: 'test'
});

      ↓ ↓ ↓ ↓ ↓ ↓

// @ts-ignore
// eslint-disable-next-line import/no-unresolved
import { createApp } from '@tramvai/core';
import { Container } from '@tinkoff/dippy';
var di = new Container();
createApp({
  name: 'test',
});


`;
